clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -156,13 +161,27 @@ public void ensureValid() {
     }
 
     /**
-     * Get the timestamp of the first record in this batch. It is always the create time of the record even if the
+     * Gets the base timestamp of the batch which is used to calculate the timestamp deltas.
+     * 
+     * @return The base timestamp or
+     *         {@link RecordBatch#NO_TIMESTAMP} if the batch is empty
+     */
+    public long baseTimestamp() {
+        return buffer.getLong(FIRST_TIMESTAMP_OFFSET);
+    }
+
+    /**
+     * Get the timestamp of the first record in this batch. It is usually the create time of the record even if the
      * timestamp type of the batch is log append time.
-     *
-     * @return The first timestamp or {@link RecordBatch#NO_TIMESTAMP} if the batch is empty
+     * 
+     * @return The first timestamp if a record has been appended, unless the delete horizon has been set
+     *         {@link RecordBatch#NO_TIMESTAMP} if the batch is empty or if the delete horizon is set
      */
     public long firstTimestamp() {
```

Comment:
As far as I can tell, the only use of this method is in tests (after we fix `DefaultRecordBatch.RecordIterator`). Maybe we can remove it or give it default access? Otherwise, the implementation is a little dangerous because it does not accurately return the first timestamp in all cases.

Reply:
I've removed the firstTimestamp() method, which included some modifications to the baseTimestamp() method and those test changes.